Failing postsubmit test: 'FormatException: No value was provided for required key: out_dir_shared'

### Type of Request

bug

### Infrastructure Environment

Postsubmit tests (LUCI)

### What is happening?

<details>
<summary>Error logs</summary>

```
05:25 +30: test/integration.shard/isolated/native_assets_test.dart: flutter build apk succeeds without libraries
05:38 +30 -1: test/integration.shard/isolated/native_assets_test.dart: flutter build apk succeeds without libraries [E]
  Exception: flutter build failed: 1
  Unhandled exception:
  FormatException: No value was provided for required key: out_dir_shared
  #0      MapJsonUtils.get (package:native_assets_cli/src/json_utils.dart:50:7)
  #1      MapJsonUtils.path (package:native_assets_cli/src/json_utils.dart:25:48)
  #2      new HookConfig (package:native_assets_cli/src/config.dart:88:38)
  #3      new BuildConfig (package:native_assets_cli/src/config.dart)
  #4      build (package:native_assets_cli/src/api/build.dart:99:18)
  #5      main (file:///Volumes/Work/s/w/ir/x/t/UUbYLX/package_with_native_assets/hook/build.dart:4:9)
  #6      _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:295:33)
  #7      _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:184:12)
  Building native assets for package:package_with_native_assets failed.
  file:///Volumes/Work/s/w/ir/x/t/UUbYLX/package_with_native_assets/hook/build.dart returned with exit code: 255.
  To reproduce run:
  /Volumes/Work/s/w/ir/x/w/flutter/bin/dart --packages=/Volumes/Work/s/w/ir/x/t/UUbYLX/package_with_native_assets/.dart_tool/package_config.json /Volumes/Work/s/w/ir/x/t/UUbYLX/package_with_native_assets/.dart_tool/native_assets_builder/545fcee0f93c80dfaac63b71612fc26c/hook.dill --config=/Volumes/Work/s/w/ir/x/t/UUbYLX/package_with_native_assets/.dart_tool/native_assets_builder/545fcee0f93c80dfaac63b71612fc26c/config.json
  stderr:
  Unhandled exception:
  FormatException: No value was provided for required key: out_dir_shared
  #0      MapJsonUtils.get (package:native_assets_cli/src/json_utils.dart:50:7)
  #1      MapJsonUtils.path (package:native_assets_cli/src/json_utils.dart:25:48)
  #2      new HookConfig (package:native_assets_cli/src/config.dart:88:38)
  #3      new BuildConfig (package:native_assets_cli/src/config.dart)
  #4      build (package:native_assets_cli/src/api/build.dart:99:18)
  #5      main (file:///Volumes/Work/s/w/ir/x/t/UUbYLX/package_with_native_assets/hook/build.dart:4:9)
  #6      _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:295:33)
  #7      _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:184:12)
  
  stdout:
  
  
  Target native_assets failed: Error: Building native assets failed. See the logs for more details.
  
  
  FAILURE: Build failed with an exception.
  
  * What went wrong:
  Execution failed for task ':app:compileFlutterBuildDebug'.
  > Process 'command '/Volumes/Work/s/w/ir/x/w/flutter/bin/flutter'' finished with non-zero exit value 1
  
  * Try:
  > Run with --stacktrace option to get the stack trace.
  > Run with --info or --debug option to get more log output.
  > Run with --scan to get full insights.
  > Get more help at https://help.gradle.org.
  
  BUILD FAILED in 9s
  Gradle task assembleDebug failed with exit code 1
  
  Resolving dependencies...
  Downloading packages...
    async 2.11.0 (2.12.0 available)
    boolean_selector 2.1.1 (2.1.2 available)
    characters 1.3.0 (1.3.1 available)
    clock 1.1.1 (1.1.2 available)
    collection 1.19.0 (1.19.1 available)
    fake_async 1.3.1 (1.3.2 available)
    matcher 0.12.16+1 (0.12.17 available)
    material_color_utilities 0.11.1 (0.12.0 available)
    meta 1.15.0 (1.16.0 available)
    path 1.9.0 (1.9.1 available)
    string_scanner 1.3.0 (1.4.0 available)
    vm_service 14.3.0 (14.3.1 available)
  Got dependencies!
  12 packages have newer versions incompatible with dependency constraints.
  Try `flutter pub outdated` for more information.
  
  Running Gradle task 'assembleDebug'...                          
  Running Gradle task 'assembleDebug'...                             10.2s
  
  test/integration.shard/isolated/native_assets_test.dart 241:11       main.<fn>.<fn>
  ===== asynchronous gap ===========================
  test/integration.shard/isolated/native_assets_test_utils.dart 275:5  inTempDir
  ===== asynchronous gap ===========================
  test/integration.shard/isolated/native_assets_test.dart 227:7        main.<fn>
```
</details>

### Steps to reproduce

_No response_

### Expected results

_No response_

--- a/packages/flutter_tools/test/integration.shard/isolated/native_assets_test.dart
+++ b/packages/flutter_tools/test/integration.shard/isolated/native_assets_test.dart
@@ -223,26 +223,6 @@ void main() {
       });
     }
 
-    testWithoutContext('flutter build $buildSubcommand succeeds without libraries', () async {
-      await inTempDir((Directory tempDirectory) async {
-        final Directory projectDirectory = await createTestProjectWithNoCBuild(packageName, tempDirectory);
-
-        final ProcessResult result = processManager.runSync(
-          <String>[
-            flutterBin,
-            'build',
-            buildSubcommand,
-            '--debug',
-            if (buildSubcommand == 'ios') '--no-codesign',
-          ],
-          workingDirectory: projectDirectory.path,
-        );
-        if (result.exitCode != 0) {
-          throw Exception('flutter build failed: ${result.exitCode}\n${result.stderr}\n${result.stdout}');
-        }
-      });
-    }, skip: true); // https://github.com/flutter/flutter/issues/158120
-
     // This could be an hermetic unit test if the native_assets_builder
     // could mock process runs and file system.
     // https://github.com/dart-lang/native/issues/90.

--- a/packages/flutter_tools/test/integration.shard/isolated/native_assets_test_utils.dart
+++ b/packages/flutter_tools/test/integration.shard/isolated/native_assets_test_utils.dart
@@ -57,61 +57,6 @@ Future<Directory> createTestProject(String packageName, Directory tempDirectory)
   return packageDirectory;
 }
 
-Future<Directory> createTestProjectWithNoCBuild(String packageName, Directory tempDirectory) async {
-  final ProcessResult result = processManager.runSync(
-    <String>[
-      flutterBin,
-      'create',
-      '--no-pub',
-      packageName,
-    ],
-    workingDirectory: tempDirectory.path,
-  );
-  if (result.exitCode != 0) {
-    throw Exception(
-      'flutter create failed: ${result.exitCode}\n${result.stderr}\n${result.stdout}',
-    );
-  }
-
-  final Directory packageDirectory = tempDirectory.childDirectory(packageName);
-
-  final ProcessResult result2 = await processManager.run(
-    <String>[
-      flutterBin,
-      'pub',
-      'add',
-      'native_assets_cli',
-    ],
-    workingDirectory: packageDirectory.path,
-  );
-  expect(result2, const ProcessResultMatcher());
-
-  await pinDependencies(packageDirectory.childFile('pubspec.yaml'));
-
-  final ProcessResult result3 = await processManager.run(
-    <String>[
-      flutterBin,
-      'pub',
-      'get',
-    ],
-    workingDirectory: packageDirectory.path,
-  );
-  expect(result3, const ProcessResultMatcher());
-
-  // Add build hook that does nothing to the package.
-  final File buildHook = packageDirectory.childDirectory('hook').childFile('build.dart');
-  buildHook.createSync(recursive: true);
-  buildHook.writeAsStringSync('''
-import 'package:native_assets_cli/native_assets_cli.dart';
-
-void main(List<String> args) async {
-  await build(args, (config, output) async {});
-}
-''');
-
-  return packageDirectory;
-}
-
 Future<void> addLinkHookDependency(String packageName, Directory packageDirectory) async {
   final Directory flutterDirectory = fileSystem.currentDirectory.parent.parent;
   final Directory linkHookDirectory = flutterDirectory

--- a/packages/flutter_tools/test/integration.shard/isolated/native_assets_without_cbuild_assemble_test.dart
+++ b/packages/flutter_tools/test/integration.shard/isolated/native_assets_without_cbuild_assemble_test.dart
@@ -1,0 +1,144 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:io' as io;
+
+import 'package:file/file.dart';
+import 'package:process/process.dart';
+import 'package:yaml/yaml.dart';
+
+import '../../src/common.dart';
+import '../test_utils.dart';
+import '../transition_test_utils.dart';
+import 'native_assets_test_utils.dart';
+
+// Regression test as part of https://github.com/flutter/flutter/pull/150742.
+void main() {
+  if (!platform.isMacOS && !platform.isLinux && !platform.isWindows) {
+    // TODO(dacoharkes): Implement Fuchsia. https://github.com/flutter/flutter/issues/129757
+    return;
+  }
+
+  const ProcessManager processManager = LocalProcessManager();
+  final String nativeAssetsCliVersionConstraint =
+      _getPackageFfiTemplatePubspecVersion();
+
+  for (final String buildCommand in <String>[
+    // Current (Host) OS.
+    platform.operatingSystem,
+
+    // On macOS, also test iOS.
+    if (platform.isMacOS) 'ios',
+
+    // On every host platform, test Android.
+    'apk',
+  ]) {
+    _testBuildCommand(
+      buildCommand: buildCommand,
+      processManager: processManager,
+      nativeAssetsCliVersionConstraint: nativeAssetsCliVersionConstraint,
+    );
+  }
+}
+
+void _testBuildCommand({
+  required String buildCommand,
+  required String nativeAssetsCliVersionConstraint,
+  required ProcessManager processManager,
+}) {
+  testWithoutContext(
+    'flutter build "$buildCommand" succeeds without libraries',
+    () async {
+      await inTempDir((Directory tempDirectory) async {
+        const String packageName = 'uses_package_native_assets_cli';
+
+        // Create a new (plain Dart SDK) project.
+        await expectLater(
+          processManager.run(
+            <String>[
+              flutterBin,
+              'create',
+              '--no-pub',
+              packageName,
+            ],
+            workingDirectory: tempDirectory.path,
+          ),
+          completion(const ProcessResultMatcher()),
+        );
+
+        final Directory packageDirectory = tempDirectory.childDirectory(
+          packageName,
+        );
+
+        // Add native_assets_cli and resolve implicitly (pub add does pub get).
+        // See https://dart.dev/tools/pub/cmd/pub-add#version-constraint.
+        await expectLater(
+          processManager.run(
+            <String>[
+              flutterBin,
+              'packages',
+              'add',
+              'native_assets_cli:$nativeAssetsCliVersionConstraint',
+            ],
+            workingDirectory: packageDirectory.path,
+          ),
+          completion(const ProcessResultMatcher()),
+        );
+
+        // Add a build hook that does nothing to the package.
+        packageDirectory.childDirectory('hook').childFile('build.dart')
+          ..createSync(recursive: true)
+          ..writeAsStringSync('''
+import 'package:native_assets_cli/native_assets_cli.dart';
+
+void main(List<String> args) async {
+  await build(args, (config, output) async {});
+}
+''');
+
+        // Try building.
+        await expectLater(
+          processManager.run(
+            <String>[
+              flutterBin,
+              'build',
+              'macos',
+              '--debug',
+            ],
+            workingDirectory: packageDirectory.path,
+          ),
+          completion(const ProcessResultMatcher()),
+        );
+      });
+    },
+  );
+}
+
+/// Reads `templates/package_ffi/pubspec.yaml.tmpl` to use the package version.
+///
+/// For example, if the template would output:
+/// ```yaml
+/// dependencies:
+///   native_assets_cli: ^0.8.0
+/// ```
+///
+/// ... theh this function would return `'^0.8.0'`.
+String _getPackageFfiTemplatePubspecVersion() {
+  final String path = Context().join(
+    getFlutterRoot(),
+    'packages',
+    'flutter_tools',
+    'templates',
+    'package_ffi',
+    'pubspec.yaml.tmpl',
+  );
+  final YamlDocument yaml = loadYamlDocument(
+    io.File(path).readAsStringSync(),
+    sourceUrl: Uri.parse(path),
+  );
+  final YamlMap rootNode = yaml.contents as YamlMap;
+  final YamlMap dependencies = rootNode.nodes['dependencies']! as YamlMap;
+  final String version = dependencies['native_assets_cli']! as String;
+  return version;
+}

--- a/packages/flutter_tools/test/integration.shard/isolated/native_assets_without_cbuild_assemble_test.dart
+++ b/packages/flutter_tools/test/integration.shard/isolated/native_assets_without_cbuild_assemble_test.dart
@@ -32,9 +32,18 @@ void main() {
     // TODO(dacoharkes): Implement Fuchsia. https://github.com/flutter/flutter/issues/129757
     return;
   }
+  
 
   const ProcessManager processManager = LocalProcessManager();
   final String constraint = _getPackageFfiTemplatePubspecVersion();
+
+  setUpAll(() {
+    processManager.runSync(<String>[
+      flutterBin,
+      'config',
+      '--enable-native-assets',
+    ]);
+  });
 
   // Test building a host, iOS, and APK (Android) target where possible.
   for (final String buildCommand in <String>[

--- a/packages/flutter_tools/test/integration.shard/isolated/native_assets_without_cbuild_assemble_test.dart
+++ b/packages/flutter_tools/test/integration.shard/isolated/native_assets_without_cbuild_assemble_test.dart
@@ -13,7 +13,20 @@ import '../test_utils.dart';
 import '../transition_test_utils.dart';
 import 'native_assets_test_utils.dart';
 
-// Regression test as part of https://github.com/flutter/flutter/pull/150742.
+/// Regression test as part of https://github.com/flutter/flutter/pull/150742.
+///
+/// Previously, creating a new (blank, i.e. from `flutter create`) Flutter
+/// project, adding `native_assets_cli`, and adding an otherwise valid build hook
+/// (`/hook/build.dart`) would fail to build due to the accompanying shell script
+/// (at least on macOS) assuming the glob would find at least one output.
+///
+/// This test verifies that a blank Flutter project with native_assets_cli can
+/// build, and does so across all of the host platform and target platform
+/// combinations that could trigger this error.
+///
+/// The version of `native_assets_cli` is derived from the template used by
+/// `flutter create --type=pacakges_ffi`. See
+/// [_getPackageFfiTemplatePubspecVersion].
 void main() {
   if (!platform.isMacOS && !platform.isLinux && !platform.isWindows) {
     // TODO(dacoharkes): Implement Fuchsia. https://github.com/flutter/flutter/issues/129757
@@ -21,9 +34,9 @@ void main() {
   }
 
   const ProcessManager processManager = LocalProcessManager();
-  final String nativeAssetsCliVersionConstraint =
-      _getPackageFfiTemplatePubspecVersion();
+  final String constraint = _getPackageFfiTemplatePubspecVersion();
 
+  // Test building a host, iOS, and APK (Android) target where possible.
   for (final String buildCommand in <String>[
     // Current (Host) OS.
     platform.operatingSystem,
@@ -37,7 +50,7 @@ void main() {
     _testBuildCommand(
       buildCommand: buildCommand,
       processManager: processManager,
-      nativeAssetsCliVersionConstraint: nativeAssetsCliVersionConstraint,
+      nativeAssetsCliVersionConstraint: constraint,
     );
   }
 }

--- a/packages/flutter_tools/test/integration.shard/isolated/native_assets_without_cbuild_assemble_test.dart
+++ b/packages/flutter_tools/test/integration.shard/isolated/native_assets_without_cbuild_assemble_test.dart
@@ -51,6 +51,7 @@ void main() {
       buildCommand: buildCommand,
       processManager: processManager,
       nativeAssetsCliVersionConstraint: constraint,
+      codeSign: buildCommand != 'ios',
     );
   }
 }
@@ -59,6 +60,7 @@ void _testBuildCommand({
   required String buildCommand,
   required String nativeAssetsCliVersionConstraint,
   required ProcessManager processManager,
+  required bool codeSign,
 }) {
   testWithoutContext(
     'flutter build "$buildCommand" succeeds without libraries',
@@ -116,8 +118,9 @@ void main(List<String> args) async {
             <String>[
               flutterBin,
               'build',
-              'macos',
+              buildCommand,
               '--debug',
+              if (!codeSign) '--no-codesign',
             ],
             workingDirectory: packageDirectory.path,
           ),

--- a/packages/flutter_tools/test/integration.shard/isolated/native_assets_without_cbuild_assemble_test.dart
+++ b/packages/flutter_tools/test/integration.shard/isolated/native_assets_without_cbuild_assemble_test.dart
@@ -136,7 +136,7 @@ void main(List<String> args) async {
 ///   native_assets_cli: ^0.8.0
 /// ```
 ///
-/// ... theh this function would return `'^0.8.0'`.
+/// ... then this function would return `'^0.8.0'`.
 String _getPackageFfiTemplatePubspecVersion() {
   final String path = Context().join(
     getFlutterRoot(),

--- a/packages/flutter_tools/test/integration.shard/isolated/native_assets_without_cbuild_assemble_test.dart
+++ b/packages/flutter_tools/test/integration.shard/isolated/native_assets_without_cbuild_assemble_test.dart
@@ -32,7 +32,7 @@ void main() {
     // TODO(dacoharkes): Implement Fuchsia. https://github.com/flutter/flutter/issues/129757
     return;
   }
-  
+
 
   const ProcessManager processManager = LocalProcessManager();
   final String constraint = _getPackageFfiTemplatePubspecVersion();

